x/ibc/04-channel/types/msgs.go

```diff
@@ -389,3 +390,169 @@ func (msg MsgChannelCloseConfirm) GetSignBytes() []byte {
 func (msg MsgChannelCloseConfirm) GetSigners() []sdk.AccAddress {
 	return []sdk.AccAddress{msg.Signer}
 }
+
+// MsgPacket receives incoming IBC packet
+type MsgPacket struct {
+	Packet      `json:"packet" yaml:"packet"`
+	Proof       commitment.ProofI `json:"proof" yaml:"proof"`
+	ProofHeight uint64            `json:"proof_height" yaml:"proof_height"`
+	Signer      sdk.AccAddress    `json:"signer" yaml:"signer"`
+}
+
+var _ sdk.Msg = MsgPacket{}
+
+// NewMsgPacket constructs new MsgPacket
+func NewMsgPacket(packet Packet, proof commitment.ProofI, proofHeight uint64, signer sdk.AccAddress) MsgPacket {
+	return MsgPacket{
+		Packet:      packet,
+		Proof:       proof,
+		ProofHeight: proofHeight,
+		Signer:      signer,
+	}
+}
+
+// Route implements sdk.Msg
+func (msg MsgPacket) Route() string {
+	return msg.DestinationPort
+}
+
+// ValidateBasic implements sdk.Msg
+func (msg MsgPacket) ValidateBasic() error {
+	if msg.Proof == nil {
+		return sdkerrors.Wrap(commitment.ErrInvalidProof, "cannot submit an empty proof")
+	}
+	if err := msg.Proof.ValidateBasic(); err != nil {
+		return sdkerrors.Wrap(err, "proof ack cannot be nil")
+	}
+	if msg.ProofHeight == 0 {
+		return sdkerrors.Wrap(ibctypes.ErrInvalidHeight, "proof height must be > 0")
+	}
+	if msg.Signer.Empty() {
+		return sdkerrors.ErrInvalidAddress
+	}
+
+	return msg.Packet.ValidateBasic()
+}
+
+// GetSignBytes implements sdk.Msg
+func (msg MsgPacket) GetSignBytes() []byte {
+	return sdk.MustSortJSON(SubModuleCdc.MustMarshalJSON(msg))
+}
+
+// GetSigners implements sdk.Msg
+func (msg MsgPacket) GetSigners() []sdk.AccAddress {
+	return []sdk.AccAddress{msg.Signer}
+}
+
+var _ sdk.Msg = MsgTimeout{}
+
+// MsgTimeout receives timeouted packet
+type MsgTimeout struct {
+	Packet           `json:"packet" yaml:"packet"`
+	NextSequenceRecv uint64            `json:"next_sequence_recv,omitempty" yaml:"next_sequence_recv,omitempty"`
+	Proof            commitment.ProofI `json:"proof" yaml:"proof"`
+	ProofHeight      uint64            `json:"proof_height" yaml:"proof_height"`
+	Signer           sdk.AccAddress    `json:"signer" yaml:"signer"`
+}
+
+// NewMsgTimeout constructs new MsgTimeout
+func NewMsgTimeout(packet Packet, nextSequenceRecv uint64, proof commitment.ProofI, proofHeight uint64, signer sdk.AccAddress) MsgTimeout {
+	return MsgTimeout{
+		Packet:           packet,
+		NextSequenceRecv: nextSequenceRecv,
+		Proof:            proof,
+		ProofHeight:      proofHeight,
+		Signer:           signer,
+	}
+}
+
+// Route implements sdk.Msg
+func (msg MsgTimeout) Route() string {
+	return msg.SourcePort
+}
+
+// ValidateBasic implements sdk.Msg
+func (msg MsgTimeout) ValidateBasic() error {
```

Comment:
is `NextSequenceRecv` optional? if so add a comment why

Reply:
It just doesn't need to be validated in this function